main/src/ca/uwaterloo/flix/language/phase/Redundancy.scala

```diff
@@ -106,13 +106,40 @@ object Redundancy extends Phase[TypedAst.Root, TypedAst.Root] {
     }
 
     // Compute the used symbols inside the definition.
-    val usedExp = visitExp(defn.exp, Env.of(defn.fparams.map(_.sym)))
+    val recursionContext = RecursionContext(defn.sym, arity(defn))
+    val usedExp = visitExp(defn.exp, Env.of(Some(recursionContext)) ++ defn.fparams.map(_.sym))
 
     // Check for unused parameters and remove all variable symbols.
-    val usedAll = (usedExp ++ checkUnusedFormalParameters(usedExp) ++ checkUnusedTypeParameters(usedExp)).copy(varSyms = Set.empty)
+    val usedAll = (usedExp and checkUnusedFormalParameters(usedExp) and checkUnusedTypeParameters(usedExp)).copy(varSyms = Set.empty)
+
+    val usedAllWithUnconditionalRecursions = if (usedAll.unconditionallyRecurses) usedAll + UnconditionalRecursion(defn.sym) else usedAll
 
     // Check if the used symbols contains holes. If so, strip out all error messages.
-    if (usedAll.holeSyms.isEmpty) usedAll else usedAll.copy(errors = Set.empty)
+    if (usedAllWithUnconditionalRecursions.holeSyms.isEmpty)
+      usedAllWithUnconditionalRecursions
+    else
+      usedAllWithUnconditionalRecursions.copy(errors = Set.empty)
+  }
+
+  /**
+    * Finds the arity of the Def, including lambdas nested directly beneath it.
+    */
+  private def arity(defn: Def): Int = {
+    val topArity = defn.fparams.size
+    val lambdaArity = defn.exp match {
+      case lambda: Expression.Lambda => arity(lambda)
+      case _ => 0
+    }
+    topArity + lambdaArity
+  }
+
+  /**
+    * Finds the arity of the Lambda: the number of lambdas nested below, including this one.
+    */
+  @tailrec
+  private def arity(lambda: Expression.Lambda, acc: Int = 0): Int = lambda.exp match {
```

Comment:
If you have an expression you can get its type and then use it to determine how many arguments are there. This is a more robust solution. Let me know if you need help understanding the structure of types -- because they are bit more complicated than you might expect.

Reply:
I wasn't able to use `Arrow` `length` directly, since it appears always to be 2 with the defs I tested, regardless of the true arity. A recursive traversal of the type seems to work, however.

I also noticed there is a [TODO](https://github.com/flix/flix/blob/ac1f20654e462dd008fe1b5dad96fcc8006c697c/main/src/ca/uwaterloo/flix/language/ast/TypedAst.scala#L39) suggesting removing tpe from the Def case class. Should I then not be relying on this field?

Reply:
Yeah. Hmmm. You can use it for now.